processor/metric/payload.go

```diff
@@ -0,0 +1,217 @@
+package metric
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/elastic/apm-server/config"
+	m "github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/utility"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var (
+	processorEntry = common.MapStr{"name": processorName, "event": docType}
+)
+
+type Sample interface {
+	Name() string
+	Value() common.MapStr
+}
+
+type Counter struct {
+	name  string
+	count float64
+	unit  *string
+}
+
+func (c *Counter) Name() string {
+	return c.name
+}
+
+func (c *Counter) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "counter",
+		"value": c.count,
+	}
+	if c.unit != nil {
+		v["unit"] = c.unit
+	}
+	return v
+}
+
+type Gauge struct {
+	name  string
+	value float64
+	unit  *string
+}
+
+func (g *Gauge) Name() string {
+	return g.name
+}
+
+func (g *Gauge) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "gauge",
+		"value": g.value,
+	}
+	if g.unit != nil {
+		v["unit"] = g.unit
+	}
+	return v
+}
+
+func CounterFrom(name string, sample map[string]interface{}) *Counter {
+	v, ok := sample["count"]
+	if !ok {
+		return nil
+	}
+	n, ok := v.(json.Number)
+	if !ok {
+		return nil
+	}
+	var unit *string
+	if u, ok := sample["unit"]; ok {
+		s := u.(string)
+		unit = &s
+	}
+	if f, err := n.Float64(); err == nil {
+		return &Counter{
+			name:  name,
+			count: f,
+			unit:  unit,
+		}
+	}
+	return nil
+}
+
+func GaugeFrom(name string, sample map[string]interface{}) *Gauge {
+	v, ok := sample["value"]
+	if !ok {
+		return nil
+	}
+	n, ok := v.(json.Number)
+	if !ok {
+		return nil
+	}
+	var unit *string
+	if u, ok := sample["unit"]; ok {
+		s := u.(string)
+		unit = &s
+	}
+
+	if f, err := n.Float64(); err == nil {
+		return &Gauge{
+			name:  name,
+			value: f,
+			unit:  unit,
+		}
+	}
+	return nil
+}
+
+type Metric struct {
+	Samples   []Sample
+	Tags      common.MapStr
+	Timestamp time.Time
+}
+
+type Payload struct {
+	Process *m.Process
+	Service m.Service
+	System  *m.System
+	Metrics []*Metric
+}
+
+func decodeSamples(raw map[string]interface{}) []Sample {
+	sl := make([]Sample, 0)
+
+	samples := raw["samples"].(map[string]interface{})
+	for name, s := range samples {
+		var sample Sample
+		sampleMap := s.(map[string]interface{})
+		sampleType, ok := sampleMap["type"]
+		if !ok {
+			// json schema should prevent this
+			logp.NewLogger("transform").Debugf("unexpected sample: %s: %s", name, s)
+		}
+		switch sampleType {
+		case "counter":
+			if c := CounterFrom(name, sampleMap); c != nil {
+				sample = c
+			}
+		case "gauge":
+			if g := GaugeFrom(name, sampleMap); g != nil {
+				sample = g
+			}
+		}
+		if sample != nil {
```

Comment:
iirc we don't have this kind of behaviour for other types, right? we either accept it all or reject all if 1 item in the list is wrong (i think)

I personally prefer the way here, but we should be consistent imo

Reply:
@simitt @roncohen thoughts on this?  We require one of the known types via json schema so the entry to this would either be a mistake there or programmatic usage of this api.  I think individual metrics should be ignored rather than failing the whole set of metrics but I agree with @jalvz that consistency with the other endpoints is important.